addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/WebCacheDeceptionScanRule.java

```diff
@@ -0,0 +1,246 @@
+                sendAndReceive(authorisedMessage);
+                String originalResponse = authorisedMessage.getResponseBody().toString();
+                URI uri = authorisedMessage.getRequestHeader().getURI();
+                ArrayList<String> extensions = new ArrayList<>();
+                String path = uri.getPath();
+                // checks whether the page with appended path gets cached or not
+                for (String ext : TEST_EXTENSIONS) {
+                    String newPath = path + "/test." + ext;
+                    uri.setPath(newPath);
+                    authorisedMessage.getRequestHeader().setURI(uri);
+                    sendAndReceive(authorisedMessage);
+                    String cachedResponse = authorisedMessage.getResponseBody().toString();
+                    String unauthorisedCachedResponse =
+                            makeUnauthorisedRequest(
+                                    uri, authorisedMessage.getRequestHeader().getMethod());
+
+                    if (checkSimilarity(unauthorisedCachedResponse, cachedResponse)) {
+                        extensions.add(ext);
+                    }
+                }
+                uri.setPath(path);
+                authorisedMessage.getRequestHeader().setURI(uri);
+                sendAndReceive(authorisedMessage);
+                if (!extensions.isEmpty()) {
+                    StringBuilder attack = new StringBuilder();
+                    for (String ext : extensions) {
+                        attack.append("/test.").append(ext).append(",");
+                    }
+                    newAlert()
+                            .setConfidence(Alert.CONFIDENCE_MEDIUM)
+                            .setMessage(authorisedMessage)
+                            .setOtherInfo(
+                                    Constant.messages.getString(
+                                            "ascanalpha.webCacheDeception.otherinfo"))
+                            .setAttack(attack.toString())
+                            .raise();
+                }
+            } catch (IOException | InterruptedException | URISyntaxException e) {
+                LOG.warn(e.getMessage(), e);
+            }
+        }
+    }
+
+    private boolean checkSimilarity(String a, String b) {
+        LevenshteinDistance distance = new LevenshteinDistance();
+        int levenshteinDistance = distance.apply(a, b);
+        // if response length is less than threshold
+        if (b.length() <= LEVENSHTEIN_THRESHOLD) {
+            return levenshteinDistance < (int) (b.length() * 0.90);
+        } else {
+            return levenshteinDistance < LEVENSHTEIN_THRESHOLD;
+        }
+    }
+
+    private String makeUnauthorisedRequest(URI uri, String method)
+            throws IOException, InterruptedException, URISyntaxException {
+        java.net.URI newURI = new java.net.URI(uri.getURI());
+        URL url = newURI.toURL();
+        CloseableHttpClient client = HttpClients.createDefault();
```

Comment:
Can a proxy be set for the client so that the request is made through ZAP?

Reply:
That might lead to the same problem (e.g. forced user would authenticate the request).

The scan rule can notify that the request was sent though.

Reply:
Actually, might as well create a new `HttpSender` and send with that, it can bypass the `HttpSenderListener`s (i.e. `sendAndReceive(HttpMessage, HttpRequestConfig)`.

Reply:
Thanks @thc202